go/genkit/session.go

```diff
@@ -0,0 +1,298 @@
+	sessionData, err := store.Get(sessionId)
+	if err != nil {
+		return nil, err
+	}
+
+	session = &Session{
+		ID:          sessionId,
+		SessionData: sessionData,
+		Store:       store,
+	}
+
+	return session, nil
+}
+
+// UpdateState takes any data and sets it as state in the store.
+func (s *Session) UpdateState(state any) error {
+	var err error
+
+	// Ensure sessionData is up to date
+	sessionData, err := s.Store.Get(s.ID)
+	if err != nil {
+		return err
+	}
+	s.SessionData = sessionData
+
+	s.SessionData.StateSchema, s.SessionData.State, err = getSchemaAndDefaults(state)
+	if err != nil {
+		return err
+	}
+
+	return s.Store.Save(s.ID, s.SessionData)
+}
+
+// UpdateMessages takes a threadName and a slice of messages.
+func (s *Session) UpdateMessages(threadName string, messages []*ai.Message) error {
+	// Ensure sessionData is up to date
+	sessionData, err := s.Store.Get(s.ID)
+	if err != nil {
+		return err
+	}
+	s.SessionData = sessionData
+
+	s.SessionData.Threads[threadName] = messages
+	return s.Store.Save(s.ID, s.SessionData)
+}
+
+// Set current session ID in context.
+func (s *Session) SetContext(ctx context.Context) context.Context {
+	ctx = context.WithValue(ctx, currentSessionID, s.ID)
+	ctx = context.WithValue(ctx, currentSessionStore, s.Store)
+
+	return ctx
+}
+
+// Find current session ID in context, returns session if found.
+func SessionFromContext(ctx context.Context) (session *Session, err error) {
+	sessionID, ok := ctx.Value(currentSessionID).(string)
+	if !ok {
+		return nil, errors.New("no session ID found in context")
+	}
+
+	store, ok := ctx.Value(currentSessionStore).(SessionStore)
+	if !ok {
+		return nil, errors.New("no session store found in context")
+	}
+
+	session, err = LoadSession(ctx, sessionID, store)
+	if err != nil {
+		return nil, err
+	}
+
+	return session, nil
+}
+
+// WithSessionID sets the session id.
+func WithSessionID(id string) SessionOption {
+	return func(s *Session) error {
+		if s.ID != "" {
+			return errors.New("cannot set session id (WithSessionID) more than once")
+		}
+		s.ID = id
+		return nil
+	}
+}
+
+// WithSessionData sets the session data.
+func WithSessionData(data SessionData) SessionOption {
+	return func(s *Session) error {
+		if s.SessionData.Threads != nil {
+			return errors.New("cannot set session data (WithSessionData) more than once")
+		}
+		s.SessionData = data
+		return nil
+	}
+}
+
+// WithSessionStore sets a session store for the session.
+func WithSessionStore(store SessionStore) SessionOption {
+	return func(s *Session) error {
+		if s.Store != nil {
+			return errors.New("cannot set session store (WithSessionStore) more than once")
+		}
+		s.Store = store
+		return nil
+	}
+}
+
+// WithStateType uses the type provided to derive the state schema.
+// If passing eg. a struct with values, the struct definition will serve as the schema, the values will serve as defaults.
+func WithStateType(state any) SessionOption {
```

Comment:
More comments below but I'm leaning towards making this very opinionated, like structs only. Unlike other situations like this in other APIs, state is de facto a map/struct, and since we need to know how to marshal it, I'm leaning towards a struct.